File: frameworks/node/express/UnitTests/compressed.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";

import request from "supertest";

import { app, expected } from "./app.ts";

const get = (size: string, encoding: string) =>
  request(app).get(`/compressed/${size}`).set({ "accept-encoding": encoding, "cache-control": "no-cache" });

// rb:test compressed.gzip_large
test("compressed.gzip_large: a large body is gzipped when asked", async () => {
  const response = await get("large", "gzip");

  // supertest inflates the body and keeps the header that says it was gzipped.
  assert.equal(response.headers["content-encoding"], "gzip");
  assert.deepEqual(response.body, expected.json("items.large.json"));
});

// rb:test compressed.gzip_small
test("compressed.gzip_small: a body under the threshold goes out as it is", async () => {
  const response = await get("small", "gzip");

  assert.equal(response.headers["content-encoding"], undefined);
  assert.deepEqual(response.body, expected.json("items.small.json"));
});

// rb:test compressed.identity_small,compressed.identity_large
for (const [id, size] of [["compressed.identity_small", "small"], ["compressed.identity_large", "large"]] as const) {
  test(`${id}: identity is answered as it is, and the handler runs every time`, async () => {
    const first = await get(size, "identity");
    const second = await get(size, "identity");

    assert.equal(second.headers["content-encoding"], undefined);
    assert.deepEqual(second.body, expected.json(`items.${size}.json`));
    assert.ok(Number(second.headers["x-rb-serial"]) > Number(first.headers["x-rb-serial"]));
  });
}

test("a route outside the family is not compressed", async () => {
  const response = await request(app).get("/json/large").set("accept-encoding", "gzip");

  assert.equal(response.headers["content-encoding"], undefined);
});

File: frameworks/go/fiber/UnitTests/compressed_test.go
package unittests

import (
	"bytes"
	"compress/gzip"
	"io"
	"net/http"
	"reflect"
	"testing"
)

// rb:test compressed.gzip_large
func TestGzipIsAnsweredWhenTheRequestAsksForIt(t *testing.T) {
	t.Run("compressed.gzip_large", func(t *testing.T) {
		before := serial(t, get(t, "/compressed/large", "Accept-Encoding", "identity"))

		response := get(t, "/compressed/large", "Accept-Encoding", "gzip", "Cache-Control", "no-cache")

		assertStatus(t, response, http.StatusOK)
		if coding := response.Header.Get("Content-Encoding"); coding != "gzip" {
			t.Fatalf("content encoding %q", coding)
		}
		reader, err := gzip.NewReader(bytes.NewReader(bodyOf(t, response)))
		if err != nil {
			t.Fatal(err)
		}
		plain, err := io.ReadAll(reader)
		if err != nil {
			t.Fatal(err)
		}
		if got, want := parsed(t, plain), payload(t, "items.large.json"); !reflect.DeepEqual(got, want) {
			t.Fatal("the gzipped body is not the payload")
		}
		if serial(t, response) <= before {
			t.Fatal("x-rb-serial did not advance")
		}
	})
}

// fasthttp compresses no body shorter than 200 bytes, and items.small is 123.
// rb:test compressed.gzip_small
func TestASmallBodyIsSentUncompressedWhenTheRequestAsksForGzip(t *testing.T) {
	t.Run("compressed.gzip_small", func(t *testing.T) {
		before := serial(t, get(t, "/compressed/small", "Accept-Encoding", "identity"))

		response := get(t, "/compressed/small", "Accept-Encoding", "gzip", "Cache-Control", "no-cache")

		if coding := response.Header.Get("Content-Encoding"); coding != "" {
			t.Fatalf("content encoding %q", coding)
		}
		assertHeader(t, response, "Vary", "Accept-Encoding")
		assertOK(t, payload(t, "items.small.json"), response)
		if serial(t, response) <= before {
			t.Fatal("x-rb-serial did not advance")
		}
	})
}

// rb:test compressed.identity_small,compressed.identity_large
func TestIdentityIsAnsweredUncompressed(t *testing.T) {
	for _, row := range []struct{ id, size string }{
		{"compressed.identity_small", "small"},
		{"compressed.identity_large", "large"},
	} {
		t.Run(row.id, func(t *testing.T) {
			response := get(t, "/compressed/"+row.size, "Accept-Encoding", "identity", "Cache-Control", "no-cache")

			if coding := response.Header.Get("Content-Encoding"); coding != "" {
				t.Fatalf("content encoding %q", coding)
			}
			assertOK(t, payload(t, "items."+row.size+".json"), response)
		})
	}
}
